lean4/src/putnam_2022_a5.lean

```diff
@@ -0,0 +1,22 @@
+import Mathlib
+
+noncomputable abbrev putnam_2022_a5_solution : ℕ := sorry
+-- 290
+/--
+Alice and Bob play a game on a board consisting of one row of 2022 consecutive squares. They take turns placing tiles that cover two adjacent squares, with Alice going first. By rule, a tile must not cover a square that is already covered by another tile. The game ends when no tile can be placed according to this rule. Alice's goal is to maximize the number of uncovered squares when the game ends; Bob's goal is to minimize it. What is the greatest number of uncovered squares that Alice can ensure at the end of the game, no matter how Bob plays?
+-/
+theorem putnam_2022_a5
+    (IsValidMove : Set (Fin 2022) → Set (Fin 2022) → Prop)
+    (IsValidMove_def : ∀ x y, IsValidMove x y ↔
+      (x = y ∧ ∀ i < 2021, i ∉ x → i + 1 ∈ x) ∨
+      ∃ i < 2021, i ∉ x ∧ i + 1 ∉ x ∧ y = x ∪ {i, i + 1})
+    (IsValidGame : List (Set (Fin 2022)) → Prop)
+    (IsValidGame_def : ∀ g, IsValidGame g ↔ (∃ gt, g = ∅ :: gt) ∧ g.Chain' IsValidMove)
+    (ConformsToStrategy : List (Set (Fin 2022)) → (Set (Fin 2022) → Set (Fin 2022)) → Prop)
+    (ConformsToStrategy_def : ∀ g s, ConformsToStrategy g s ↔
+      ∀ (i) (h : i + 1 < g.length), Even i → g[i + 1] = s g[i]) :
+    IsGreatest
+      {n | ∃ s, (∀ x, IsValidMove x (s x)) ∧ ∀ g,
+        IsValidGame g → ConformsToStrategy g s → ∃ gh x, g = gh ++ [x] → n ≤ xᶜ.ncard}
```

Comment:
I think this becomes trivially true for all n, because `∃ gh x, g = gh ++ [x] → n ≤ xᶜ.ncard` should be `∃ gh x, g = gh ++ [x] ∧ n ≤ xᶜ.ncard`, right?

Reply:
Nice catch!